src/uucore/src/lib/features/quoting_style.rs

```diff
@@ -351,30 +432,61 @@ fn escape_name_inner(name: &OsStr, style: &QuotingStyle, dirname: bool) -> Strin
             };
 
             let (escaped_str, contains_quote_chars) = if *escape {
-                shell_with_escape(&name, quotes)
+                shell_with_escape(name, quotes)
             } else {
-                shell_without_escape(&name, quotes, *show_control)
+                shell_without_escape(name, quotes, *show_control)
             };
 
-            match (must_quote | contains_quote_chars, quotes) {
-                (true, Quotes::Single) => format!("'{escaped_str}'"),
-                (true, Quotes::Double) => format!("\"{escaped_str}\""),
-                _ => escaped_str,
+            if must_quote | contains_quote_chars && quotes != Quotes::None {
+                let mut quoted_str = Vec::<u8>::with_capacity(escaped_str.len() + 2);
+                let quote = if quotes == Quotes::Single {
+                    b'\''
+                } else {
+                    b'"'
+                };
+                quoted_str.push(quote);
+                quoted_str.extend(escaped_str);
+                quoted_str.push(quote);
+                quoted_str
+            } else {
+                escaped_str
             }
         }
     }
 }
 
 /// Escape a filename with respect to the given style.
-pub fn escape_name(name: &OsStr, style: &QuotingStyle) -> String {
-    escape_name_inner(name, style, false)
+pub fn escape_name(name: &OsStr, style: &QuotingStyle) -> OsString {
+    #[cfg(unix)]
+    {
+        let name = name.as_bytes();
+        OsStringExt::from_vec(escape_name_inner(name, style, false))
+    }
+    #[cfg(not(unix))]
+    {
+        let name = name.to_string_lossy();
+        String::from_utf8_lossy(&escape_name_inner(name.as_bytes(), style, false))
+            .to_string()
+            .into()
+    }
```

Comment:
Please avoid code duplication by using `os_str_from_bytes` or `os_string_from_vec` in `uucore/src/lib/lib.rs`

Reply:
The applicable functions (`os_str_as_bytes`, `os_str_from_bytes`, `os_string_from_vec`) error on invalid UTF-8 outside of unix, rather than doing a lossy conversion (which is what we had before this PR). Is that actually what we want here? E.g., if someone mounts a non-NTFS partition on Windows, or calls printf with some random bytes, is it better to show names with replacement characters, or to fail?

Reply:
I'm not sure what's the expected behavior here, I guess using lossy conversion is good, until someone wants to put it to the test.

Anyway, the conversion logic with the `#[cfg]`s should be placed in a `os_string_from_vec_lossy` function in `lib.rs`, in case it may be used somewhere else in the codebase

Reply:
Tried implementing this, and I don't think there's a way to do it. We need something like `os_str_as_bytes_lossy`, but the return type would depend on the platform:
 - `&[u8]` on unix, or 
 - `Cow<'_, str>` (which you could then get a `&[u8]` from with `as_bytes()`) otherwise.

At that point, there's no advantage to having a single helper function. I.e., `name` in these branches has two different types, because when on unix, we can always avoid extra allocations, while on not(unix), we need an owned/CoW type to handle the new string with replacement characters.

Reply:
Ah, never mind, we can just map the CoW type manually.